Register custom icons in class macro

Paperback opened the pull request:
Thanks for the awesome work everyone puts into this project. This is my first rust contribution, so I'm keen to get some feedback and learn.

This pull request closes issue #1200.

It adds a new `icon` attribute to the `#[class]` macro.

```rust

#[derive(GodotClass)]
#[class(base = Sprite2D, icon = "res://addons/my_extension/player.svg")]
pub struct PlayerOne {}

// default icon behaviour
#[derive(GodotClass)]
#[class(base = Sprite2D)]
pub struct PlayerTwo {}
```

Because the icon attribute is a `TokenStream`, we can also write:
```rust
const MY_ICON: &str = "res://icon.svg"
#[derive(GodotClass)]
#[class(icon = MY_ICON)]
pub struct Player {}
``` 


__Considerations__
- Requires api 4.4 >= as it uses `GDExtensionClassCreationInfo4::icon_path`
- The `.gdextension` file `[icon]` section declarations seem to take precedence

Thanks!

Yarwin commented on godot-core/src/registry/class.rs:
```diff
@@ -458,6 +467,26 @@ fn fill_class_info(item: PluginItem, c: &mut ClassRegistrationInfo) {
+            let _ = icon; // mark used
+            #[cfg(since_api = "4.4")]
+            if let Some(icon_path) = icon {
+                // Convert to GString and store in global map to keep it alive for program lifetime
```
End all the comments with a dot `.`, to make it clear where they end – it allows us to assert that given comment is present in full and part of it haven't been lost in any refactor and whatnot.
 
```suggestion
                // Convert to GString and store in global map to keep it alive for program lifetime.
```

Yarwin commented on godot-core/src/registry/class.rs:
```diff
@@ -458,6 +467,26 @@ fn fill_class_info(item: PluginItem, c: &mut ClassRegistrationInfo) {
+                // even after the lock guard is dropped. We must retrieve the pointer while holding
+                // the lock to ensure the HashMap isn't being modified concurrently.
+                let icon_ptr = icon_map.get(&c.class_name).unwrap().string_sys()
+                    as sys::GDExtensionConstStringPtr;
```
it is recommended to use the `cast` method on raw pointers instead of `ptr as ...`, so as to avoid accidentally mixing up (or introducing) mutability. I know that `as` is ultra-present in the codebase, we are moving away fro mit.

https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.cast-1
https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.cast

Yarwin replied:
maybe:

```suggestion
                let icon_ptr: sys::GDExtensionConstStringPtr = icon_map.get(&c.class_name).unwrap().string_sys().as_const();
```

?

https://github.com/godot-rust/gdext/blob/93a81ccaa0e2f32a53e6be63f2148fa3a03842e5/godot-ffi/src/extras.rs#L76

Yarwin commented on godot-core/src/registry/class.rs:
```diff
@@ -559,6 +588,23 @@ fn register_class_raw(mut info: ClassRegistrationInfo) {
+
+            // SAFETY: The GString is stored in a static HashMap, so the pointer remains valid
+            // even after the lock guard is dropped. We must retrieve the pointer while holding
+            // the lock to ensure the HashMap isn't being modified concurrently.
```
as above – either make it a note or mark unsafe thing – which requires respecting safety remarks to not make it unsound – with unsafe block

Yarwin reviewed (changes requested):
First of all – thanks!!!! 

----

1) `ICON_STRINGS_BY_NAME` which keeps `GString`s is not being pruned during deinit of the library, which means memory leak during, for example, hot reload.
Please check it and report if I'm wrong. See also [`cleanup`](https://github.com/godot-rust/gdext/blob/978867c3e0cfc80352cecf33fe3bbc5443ed4860/godot-core/src/meta/mod.rs#L103)

2) This feature is not tested at all – I think I would create new file in `itest/rust/src/register_tests` and annotate it with `#![cfg(since_api = "4.4")]` (similarly to register_docs.rs). Tests allows us to watch for memory leaks, shield us against regression etc – additionally we need to guarantee that all the used APIs are OKay and give an example how they can be used.

3) I would remove `ExtensionConfig` (up to discussion) – right now its only purpose is to set default fallback icon for classes, which will be useless and counter-productive after 4.6: https://github.com/godotengine/godot/pull/108607. I'm also finding it a little confusing (like, docs mention tools and build scripts – how would I even use it there?)

4) `#[class(icon = MY_ICON)]` is not documented in `#[derive(GodotClass)]` docs.

5) Address various silly things, like `SAFETY` remarks under now-safe blocks, comments not ending with a dot etc.

Paperback commented on godot-core/src/registry/class.rs:
```diff
@@ -559,6 +588,23 @@ fn register_class_raw(mut info: ClassRegistrationInfo) {
         info.godot_params.get_virtual_func = info.user_virtual_fn.or(info.default_virtual_fn);
     }
 
+    // If no class-level icon was set, use the extension-level icon as fallback (if available)
```
In that case the gdextension icon attr would not be needed at all right, or am I misunderstanding you? That sounds easier for me 😅

At the moment, I believe the gdextension icon is useful because you can show end users of your extension what nodes are 'extension' vs 'standard' godot. This distinction, especially to plugin authors seems somewhat expected to me. If you remove the icon attr, it should revert back to inheriting base icon in 4.6- so you get both worlds right? 

Of course, there's still a pathway for that use-case as you could always set each icon yourself, so the trade-off very well might not be worth the extra complexity. I'll leave that up to your judgement I think.

Paperback commented on godot-macros/src/class/derive_godot_class.rs:
```diff
@@ -198,7 +206,7 @@ pub fn derive_godot_class(item: venial::Item) -> ParseResult<TokenStream> {
         ::godot::sys::plugin_add!(#prv::__GODOT_PLUGIN_REGISTRY; #prv::ClassPlugin::new::<#class_name>(
             #prv::PluginItem::Struct(
-                #prv::Struct::new::<#class_name>()#(.#modifiers())*
+                #prv::Struct::new::<#class_name>()#(.#modifiers())*#icon_modifier
```
I did intend to do this, but ran into a snag and decided to take the easier route, I would like to do exactly that though :)

Paperback commented on godot-core/src/registry/class.rs:
```diff
@@ -458,6 +467,26 @@ fn fill_class_info(item: PluginItem, c: &mut ClassRegistrationInfo) {
+                // Get pointer after insertion, while lock is still held
+                // SAFETY: The GString is stored in a static HashMap, so the pointer remains valid
+                // even after the lock guard is dropped. We must retrieve the pointer while holding
+                // the lock to ensure the HashMap isn't being modified concurrently.
```
Just removed some of the more wordier stuff and let the code speak for itself, if you think it's deserving of `unsafe` let me know, I don't think so considering the other global guards don't seem to.

Yarwin commented on godot-macros/src/lib.rs:
```diff
@@ -561,11 +572,22 @@ use crate::util::{bail, ident, KvParser};
+        export_group,
+        export_subgroup,
+        init,
+        icon
```
icon is not a stand-alone attribute, but part of `#[class(...)]` attribute, ergo it doesn't belong here.

Yarwin commented on itest/rust/src/register_tests/icon_test.rs:
```diff
@@ -0,0 +1,75 @@
+
+use crate::framework::itest;
+
+const REF_COUNTED_ICON: &str = "res://icons/ref_counted_icon.svg";
```
Aren't paths to all these icons invalid? Don't we check the one and the same case?

Yarwin commented on godot-core/src/meta/mod.rs:
```diff
@@ -102,4 +102,10 @@ pub(crate) use reexport_crate::*;
+    {
+        let mut icons = crate::registry::class::global_icons_by_name();
+        *icons = std::collections::HashMap::new();
+    }
```
technically right, but can't we just use a more readable `clear`? https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.clear

```suggestion
    #[cfg(since_api = "4.4")]
     crate::registry::class::global_icons_by_name().clear();
```

Yarwin commented:
> Just poking around `LoadedClass` and ClassMetadata, maybe instead of the specific icon global guard, I should use that instead? For cleanup and integration tests, I want to be able to access that lock so I've made it pub(crate) for now, but I resent that.

I think using `LoadedClass` would be the right call! We wouldn't have to worry about cleanup and whatnot, since it would be handled auto-magically while dropping the loaded class itself (`unregister_classes` does it automatically after removing `Vec` of loaded classes).

Paperback commented:
Ok, thank you both for the feedback, I've reworked this a little bit:

1. Added `ExtensionLibrary::default_icon() -> &'static str`. Can be changed to `GString`.
2. Changed it so icon string is on `LoadedClass` instead of a global lock.
3. Added `init::DEFAULT_ICON` global, resets to empty on deinitialization.
4. Simplified icon itest.

Should I add `default_icon()` to the hot-reload test? That feature is not included in tests at the moment.

Thanks!

Bromeon commented on godot-core/src/registry/class.rs:
```diff
@@ -60,6 +60,12 @@ fn global_dyn_traits_by_typeid() -> GlobalGuard<'static, HashMap<any::TypeId, Ve
 pub struct LoadedClass {
     name: ClassId,
+
+    // Class icon needs to be retained for registered for class lifetime, this is not accessed directly.
```
Grammar? Not clear what you mean.

Bromeon commented on godot-core/src/obj/traits.rs:
```diff
@@ -44,6 +44,13 @@ where
+    /// You can also set an icon via `#[class(icon = "path/to/icon.svg")]`.
+    fn icon() -> &'static str {
+        ""
+    }
```
Not 100% sure if this should be in the `GodotClass` trait... there is quite a few registration properties that aren't currently exposed like this (e.g. `#[class(internal)]`). For builder APIs, it could otherwise be a separate step:
```rs
builder.class::<MyClass>()
    .icon("path/to/icon.svg")
    .internal()
    .build()
```

---

But _if_ we decide `GodotClass` is the right place -- which needs further discussion -- then:

1. This should document how the path has to look, i.e.
   - relative to where
   - whether it should start with `res://` or not
   - maybe an example

2. Could be named `editor_icon`. Since trait methods can be called unqualified, and user-defined classes having their own `icon` isn't that unlikely.

Bromeon commented:
You accidentally force-pushed 0 commits, which auto-closed the PR. Just push again (please don't open a new PR) 🙂

You also see the Git revisions in the timeline above, in case you need to restore something.
